plenum/test/node_request/message_request/test_node_requests_missing_preprepare.py

```diff
@@ -0,0 +1,176 @@
+
+    if request.param == 'send_bad':
+        orig_method = bad_node.nodeMsgRouter.routes[MessageReq]
+
+        def send_bad(self, msg, frm):
+            if msg.msg_type == PREPREPARE:
+                resp = self.replicas[msg.params['instId']].getPrePrepare(
+                    msg.params['viewNo'], msg.params['ppSeqNo'])
+                resp = updateNamedTuple(resp, digest='11908ffq')
+                self.sendToNodes(MessageRep(**{
+                    f.MSG_TYPE.nm: msg.msg_type,
+                    f.PARAMS.nm: msg.params,
+                    f.MSG.nm: resp
+                }), names=[frm, ])
+            else:
+                return orig_method(msg, frm)
+
+        bad_node.nodeMsgRouter.routes[MessageReq] = types.MethodType(send_bad,
+                                                                     bad_node)
+        return primary_node, bad_node, good_non_primary_node, slow_node, \
+               other_nodes, send_bad, orig_method
+
+
+def test_node_requests_missing_preprepare_malicious(looper, txnPoolNodeSet,
+                                                    client1, wallet1,
+                                                    client1Connected,
+                                                    malicious_setup, teardown):
+    """
+    A node has bad network with primary and thus loses PRE-PREPARE,
+    it requests PRE-PREPARE from non-primaries once it has sufficient PREPAREs
+    but one of the non-primary does not send the PRE-PREPARE
+    """
+    # primary_node = get_master_primary_node(txnPoolNodeSet)
+    # slow_node = getNonPrimaryReplicas(txnPoolNodeSet, 0)[-1].node
+    # other_nodes = [n for n in txnPoolNodeSet if n != slow_node]
+    # bad_node = [n for n in other_nodes if n != primary_node][0]
+    # good_non_primary_node = [n for n in other_nodes if n != slow_node
+    #                          and n != bad_node and n != primary_node][0]
+    primary_node, bad_node, good_non_primary_node, slow_node, other_nodes, \
+    bad_method, orig_method = malicious_setup
+
+    slow_node.nodeIbStasher.delay(ppDelay(300, 0))
+
+    def get_reply_count_frm(node):
+        return sum([1 for entry in slow_node.spylog.getAll(
+            slow_node.process_message_rep)
+                    if entry.params['msg'].msg_type == PREPREPARE and
+                    entry.params['frm'] == node.name])
+
+    old_reply_count_from_bad_node = get_reply_count_frm(bad_node)
+    old_reply_count_from_good_node = get_reply_count_frm(good_non_primary_node)
+    old_discarded = countDiscarded(slow_node.master_replica, 'does not have '
+                                                        'expected state')
+
+    send_reqs_batches_and_get_suff_replies(looper, wallet1, client1, 10, 2)
+
+    waitNodeDataEquality(looper, slow_node, *other_nodes)
+
+    assert check_if_all_equal_in_list([n.master_replica.ordered
+                                       for n in txnPoolNodeSet])
+
+    if bad_method.__name__ == 'do_not_send':
```

Comment:
Most of the code in `malicious_setup` is conditional depending on the parameter. And there is also condition in the test itself. 
It's better to have some duplicated code in tests, then make tests more complicated and unreadable.